### DISCUSSION TITLE
Add possibility for authors to describe a "goal" or an "intent" of a kata

### DISCUSSION THREAD

**Discussion by hobovsky** [codewars/codewars.com#2588] [Ideas]
Currently, many kata are inconsistent between their translations w.r.t. difficulty, performance requirements, or even task definition. Sometimes it's not really known what a kata is about and it's difficult to translate it accurately, or deduplicate it (see [this post](https://github.com/codewars/content-issues/issues/73#issuecomment-1023183579)). Sometimes, translation just breaks a kata because new language provides built-in means of solving it, while it was originally a challenging task in its original language.

My proposal is to enable authors (or content maintainers) to specify what is actual intent of a created kata, expectations of performance, etc. For example, "Check if a number is prime with complexity better than O(n)", or "Perform bigint arithmetic in languages which have no support for bigintegers", or "Extend existing type by adding an extension method to it", etc. Such information could be used then by authors and reviewers to ensure than the main idea behind the kata is not lost between translations, and over the time.

The information could be added for example with additional field in kata editor (kata-wide, like tags, or title, or category), or if extending the model is infeasible, maybe in some other way. Maybe a comment in a description? Drawback of comments is that it would be visible to users, potentially serving as a spoiler. Maybe you have some other idea?

What do you think?

**Comment by stevemartis** [comment]
II lIke this. I have read a few horror stories about good katas getting early retirements due to drops in satisfaction ratings from bad translations, or perhaps differences in languages. If a translator knows the intent, a method from a particular language can be blocked if it is known that it conflicts with the intent of the original kata. Like you said, a public notice of intent may serve as a spoiler. 

Perhaps a message from the author that can only be seen by a prospective translator. A translation approver could also see this message, and determine if the translator respected this in their translation.

**Reply by hobovsky** [reply]
I think it could serve also an additional purpose of making authors aware that usually kata are not language-specific. Authors usually look at kata through a keyhole of their favorite language, make everything very JS- or Python-specific, and then it's difficult to, for example, avoid merge conflicts because description uses language-specific terms which need to be reworded by translators.

Maybe having such field would rise awareness of the fact that kata are not available in a single language forever.

**Comment by JohanWiltink** [comment]
My first reaction was "Isn't that what the `Description` is for?" but I guess it could - in limited, rare cases - be a spoiler.

Having it as a spoilered comment in `Discourse` would be a terrible hack, and not very effective because not easily retrievable for translators.

An optional free form field like `tag`s _might_ work, if implemented properly, _and actually used by authors._ Most, of course, would not. I have my cynical doubts if this scheme is ever going to significantly improve the current situation. Writing descriptions is already not the forte of most kata authors, and you would have them write actual documentation on what they don't even understand?

This would be a good reason to bump up the kata authoring privilege to about `100 000` honour.

**Reply by stevemartis** [reply]
In my 3kyu view...authoring honor requirements are fine. The original beta katas are thoroughly tested by the dans. Many issues are brought up in discourse that involve the kata description. The translations are the weak link it seems. 

I hesitate to suggest that honor requirements be raised for translating or approving translations as there is already a backlog of approvable katas...

Therefore, I suggest that::
1) As you have suggested, there needs to be a method to clearly let a translator and approver know the intent of the kata.

2) Translators and approvers need to be notified if they create/approve translations that are contradictory to this intent. Mods can be given the ability to temporarily revoke privileges if warnings are not heeded.

In the end this is another system that may fall apart or simply not be manageable.

**Reply by hobovsky** [reply]
> This would be a good reason to bump up the kata authoring privilege to about `100 000` honour.

I don't want this discussion to divert in direction of prerequisites for authoring, because this is another can of worms. I will start another discussion on this. But I have some ideas why changing the honor requirement is not the way to go. I will try to explain them in a dedicated thread.

> An optional free form field like tags might work, if implemented properly, and actually used by authors. Most, of course, would not.

I think this could be handled by reviewers. We have many users who are very eager to point out uneven margins of description and such, and I'd hope that if the author does not get this part right, then reviewers will be more than happy to point this out to them with a pointy stick.

**Reply by JohanWiltink** [reply]
> pointy stick

more of a blunt stick person myself ..

I'll put my thoughts on authoring requirements in over there then - no need to poison this discussion with it.

**Comment by Kacarott** [comment]
I like the idea, but I think it only really works well if there is a specific field which is hidden from users who have not solved.  Perhaps this could also be tied in with a better search? If a kata must have a plain-english direct description of what it wants to achieve, then it would be a lot easier to collect keywords from there to use when searching for katas, rather than trying to guess all the creative titles that authors give.

Then again, having those terms be searchable might actually break the spoiler thing...

**Comment by ejini6969** [comment]
The one out of the two ways to fix this issue is to fix the tagging system which is nearly impossible at this point of time and with the given limitations (total number of katas, too much classification ways, hard to distinguish between them etc.).

Another way is to add a note about complexity expected by the kata in a specified format, like a lot of myjinxin kata series which uses `NOTE: O(N) Complexity` . But I doubt this is applicable with these huge amount of katas plus the lack of understanding on documentation from author's or translator's side. (TBH: Despite the strict requirements being stated, I bet someone will still intentionally or unintentionally missed it, unless CW set it as a must-filled field in authoring or translating katas ! but this could go wrong easily as some translators are unaware of/ unable to distinguish between different time complexities which will create more chaos.

Also: no need to create separate categories for such purpose! Just put everything into description (Note section better). If translators use the wrong complexity or forgot to disable builtins or other modules or so, it is the reviewer's job to remind the translator and even reject the translation if it is unfixable or undoable within some languages like Python. So, I don't see this as a big issue now.

**Comment by JohanWiltink** [comment]
Sideways related: kata authors should document the intent of their random tests.

Reverse engineering existing random tests of any serious complexity gets old _fast,_ yet _noone,_ absolutely _noone_ ( full disclosure: me neither ) documents how \[ biased \] random tests should be built, even if code is _not_ self-explanatory. ( The same _often,_ but not always, goes for the example solution. Not the reference one but the author's solution everyone can immediately see after solving. )

If authors won't do that where they can document in secrecy to their heart's content, what would make anyone think they'll do it for the kata itself?

( Yes, I am currently translating a kata .. and reverse engineering the \<bleep\>ing random tests. )

**Comment by kazk** [comment]
I like the idea. All of our challenges at Qualified have a field named "purpose" explaining things like what the challenge is designed for, what are tested including performance requirements, and what to look for when reviewing submitted solutions, etc.

This needs to be hidden until completed to allow authors to explain without worry about spoilers. It can be shown on the kata discussions page above the rank assessment.

**Comment by Kacarott** [comment]
This suggestion seems closely related to (and perhaps solves?) #2911
